dev/buildtool/versions_commands.py

```diff
@@ -0,0 +1,432 @@
+            "alias": f"v{spinnaker_version}",
+            "changelog": f"{SPINNAKER_CHANGELOG_BASE_URL}/{spinnaker_version}-changelog/",
+            "lastUpdate": current_epoch_ms,
+            "minimumHalyardVersion": f"{minimum_halyard_version}",
+            "version": spinnaker_version,
+        }
+        return release
+
+    @staticmethod
+    def get_major_minor_patch_version(version):
+        """Split a M.m.p version string into an list of [major, minor, patch]."""
+        major, minor, patch = version.split(".")
+        return [major, minor, patch]
+
+    def add_or_replace_release(self, new_release, versions):
+        """Add new minor release versions to version list. If it's a new patch
+        release version then find the previous patch version in the list and
+        replace it. For example 1.28.0 will be added. 1.27.1 will replace 1.27.0."""
+
+        new_release_semver = new_release["version"]
+
+        patched_existing_version = False
+
+        # loop over list in versions.yml["versions"]
+        for idx, existing_release in enumerate(versions):
+
+            existing_release_semver = existing_release["version"]
+
+            major1, minor1, patch1 = self.get_major_minor_patch_version(
+                new_release_semver
+            )
+            major2, minor2, patch2 = self.get_major_minor_patch_version(
+                existing_release_semver
+            )
+
+            # check if matching minor, eg: new 1.27.1 and current version_block is 1.27.x
+            if (major1 == major2) and (minor1 == minor2):
+                if patch1 == patch2:
+                    raise_and_log_error(
+                        ValueError(
+                            f"Spinnaker version already exists in versions.yml: {new_release_semver}"
+                        )
+                    )
+
+                elif patch1 < patch2:
+                    # new version older than previous release.
+                    # e.g: spinnaker_version = 1.27.0 but versions.yml has 1.27.1
+                    raise_and_log_error(
+                        ValueError(
+                            f"Spinnaker version {new_release_semver} superceded by newer patch version: {existing_release_semver}"
+                        )
+                    )
+
+                else:
+                    logging.info(
+                        "Replacing version %s with new version %s",
+                        existing_release_semver,
+                        new_release_semver,
+                    )
+                    versions[idx] = new_release
+                    patched_existing_version = True
+
+        if not patched_existing_version:
+            versions.append(new_release)
+
+        return versions
+
+    def build(
+        self,
+        new_spinnaker_version,
+        minimum_halyard_version,
+        latest_halyard_version=None,
+    ):
+        """Build versions.yml. If an optional versions_dict is supplied then update."""
```

Comment:
I don't see a way to supply `versions_dict` here.

Reply:
```suggestion
        """Build versions.yml."""
```

Sorry the description needed to be corrected after moving `versions_dict` to a constructor.

Reply:
And it is a very descriptive description... :joy: